feat(ui): Add Suspend and Resume for Cronworkflow in UI. Closes #1975

changhc opened the pull request:
#### Screenshot
![image](https://user-images.githubusercontent.com/4591982/81425042-6fc5c300-9189-11ea-99d3-932ff1f74d40.png)

Updated:
![image](https://user-images.githubusercontent.com/4591982/81463886-94ed1c80-91ef-11ea-8e78-b8d1c27a601c.png)



Closes #1975.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [ ] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [ ] I've signed the CLA.
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to USERS.md.

simster7 commented on ui/src/app/cron-workflows/components/cron-workflow-details/cron-workflow-details.tsx:
```diff
@@ -93,4 +109,32 @@ export class CronWorkflowDetails extends BasePage<RouteComponentProps<any>, Stat
+                    type: NotificationType.Error
+                });
+            })
+            .then(() => this.getCronWorkflow());
```
Doesn't the `update` function return the new, updated Workflow? I don't think we need to make this extra API call

changhc replied:
You're right. I've updated this part. Thanks!

changhc commented:
I discovered some bugs while addressing the comment. They were mainly due to that the content in the editor wasn't updated when the cronworkflow was updated. I've already fixed this.

simster7 commented on ui/src/app/shared/components/yaml/yaml-editor.tsx:
```diff
@@ -41,7 +42,7 @@ export class YamlEditor<T> extends React.Component<Props<T>, State> {
                     />
                 ) : (
-                    <YamlViewer value={this.state.value} />
+                    <YamlViewer value={jsYaml.dump(this.props.value)} />
```
Why is this change necessary?

changhc replied:
Consider the following use case:
* A cronworkflow is running
* A user clicks `suspend` and the cronworkflow is suspended
* The user then edits the workflow and submits the updated version

Without this change, the update will fail since the suspended workflow is not reflected in the editor, i.e., `suspend: true` is missing, and an error will be returned asking the user to submit with a newer version of the cronworkflow.

Therefore, we need to use the prop value to display the updated value.

simster7 replied:
Correct, but if `this.props.value` is changed, the entire react component would re-render. I.e., the constructor would get called again and `this.state.value` would get filled with the new prop value. This is a feature of react.

Do you still think we need this change?

changhc replied:
Initially I had the same thought as yours, but it turned out that the editor didn't rerender.

simster7 replied:
In this case can we use this solution: https://stackoverflow.com/a/51309488? I'd like us to stay consistent in using `this.state.value` across the file.

changhc replied:
Updated

simster7 commented on ui/src/app/cron-workflows/components/cron-workflow-details/cron-workflow-details.tsx:
```diff
@@ -8,10 +8,12 @@ import {Loading} from '../../../shared/components/loading';
 
 interface State {
     cronWorkflow?: CronWorkflow;
+    editing: boolean;
```
I'm not too sure we need this... it would make sense to me if I was editing the YAML and _without saving my edits_ pressed Suspend/Resume that my changes would be lost.

changhc replied:
This makes sense too, but I think this is safer. The ideal design would be having an additional `cancel` button besides the `submit` button while still disabling the suspend button during editing. I think this flow is more natural and ensures safety simultaneously.

simster7 replied:
I agree that it's safer, but I can see some potential for edge case bugs. For example, in this implementation this flag is only set to false if the underlying object is changed. Does this flag work correctly if a user clicks edit then save without making any changes?

I would rather move forward without this change for now, and add it later if necessary. Let me know if you feel strongly otherwise.

changhc replied:
You're right. To solve this, we should not set `editing: false` in `onChange` but do this in another function and pass it in as a separate props. This should be called after `.then(updated => props.onChange(updated))`.

I don't feel this is troublesome or will severely affect other things but can improve UX a little bit, so to me it's still beneficial. If you still have concerns, we can skip this now.

simster7 replied:
Let's skip this for now please. You're welcome to open a separate PR with this change, though. If you do, please make sure that it's consistent across different resource kinds (`Workflow`, `WorkflowTemplate`, etc.)

simster7 commented:
@alexec Seems like CircleCI is using a dev version of Mockery? See `codegen` check

simster7 commented:
@changhc Can you merge with master to fix the CI?
